File: account-server/server_api/BUY_WITH_GOLD.py
import json
from PIL import Image, UnidentifiedImageError
from datetime import datetime
from utils import str_to_date

# BUY_WITH_GOLD
# 1. El cliente envia BUY_WITH_GOLD(session_id) al servidor.
# 2. El servidor comprueba que existe esa session_id y no esta caducada.
# 3. El servidor responde con el resource o con error.
async def BUY_WITH_GOLD(req, ws, send_message, accounts_cache, game_config):

    # Aqui se debe comprobar que la req es correcta y en caso contrario se ignora sin enviar nada al cliente.
    if not all(k in req for k in ('request_id', 'session_id', 'game_name', 'item_type', 'item_id')):
        print("ERROR: BUY_WITH_GOLD bad input parameters.")
        return

    error_response = {
        "error": "Invalid ticket.",
        "request_id": req['request_id']
    }

    valid = False

    ### Comprueba si existe esa session_id

    session_id = req["session_id"]

    query = {"session_id": session_id}
    projection = 'guest_account_id session_expire player_readonly_data'
    response = await accounts_cache.read_account(req, ws, send_message, query, projection)
    if response is None:
        return

    ### Comprueba si la session esta caducada

    if 'session_expire' in response:
        # session_expire_date = datetime_from_json_date_str(response['session_expire'])
        # if session_expire_date >= datetime.datetime.now():
        if 'guest_account_id' in response or str_to_date(response['session_expire']) >= datetime.now():

            if req['game_name'] == 'tanks':

                # SKIN

                if req['item_type'] == 'skin':
                    player_readonly_data = response.pop('player_readonly_data', None)
                    assert player_readonly_data is not None
                    max_number_of_skins = game_config['tanks']['MAX_NUMBER_OF_SKINS']
                    item_id = req['item_id']
                    if isinstance(item_id, (int, float)) and not isinstance(item_id, bool):
                        if item_id >= 0 and item_id < max_number_of_skins:                            
                            value = game_config['tanks']['SKIN_PRICES'][item_id]
                            if isinstance(value, (int, float)) and not isinstance(value, bool) and value >= 0:
                                player_readonly_data = json.loads(player_readonly_data)
                                skin_ids = player_readonly_data['skin_ids']
                                gold_total = player_readonly_data['gold']
                                if value <= gold_total:
                                    item_id = int(item_id)
                                    if item_id not in skin_ids:
                                        player_readonly_data['gold'] = gold_total - value
                                        skin_ids.append(item_id)

                                        ### Update player_readonly_data

                                        query = {"session_id": session_id}
                                        update = {}
                                        update['player_readonly_data'] = json.dumps(player_readonly_data)

                                        projection = 'session_id player_readonly_data'
                                        response = await accounts_cache.update_account(req, ws, send_message, query, update, projection)
                                        if response is not None:
                                            response["request_id"] = req['request_id']
                                            await send_message(response, ws)  # send response
                                            valid = True
                                        else:
                                            error_response["error"] = "There was a problem updating the account."
                                    else:
                                        error_response["error"] = "You already own that car.\nYou do not need to buy it again."
                                else:
                                    error_response["error"] = "Could not buy.\nYou do not have enough gold."
                            else:
                                error_response["error"] = "Value not valid."
                        else:
                            error_response["error"] = "Item id out of bounds"
                    else:
                        error_response["error"] = "Item id not valid."
                
                # FLAG

                elif req['item_type'] == 'flag':
                    player_readonly_data = response.pop('player_readonly_data', None)
                    assert player_readonly_data is not None
                    max_number_of_flags = game_config['tanks']['MAX_NUMBER_OF_FLAGS']
                    item_id = req['item_id']
                    if isinstance(item_id, (int, float)) and not isinstance(item_id, bool):
                        if item_id >= 0 and item_id < max_number_of_flags:
                            value = game_config['tanks']['FLAG_PRICE']
                            if isinstance(value, (int, float)) and not isinstance(value, bool) and value >= 0:
                                player_readonly_data = json.loads(player_readonly_data)
                                flag_ids = player_readonly_data['flag_ids']
                                gold_total = player_readonly_data['gold']
                                if value <= gold_total:
                                    item_id = int(item_id)
                                    if item_id not in flag_ids:
                                        player_readonly_data['gold'] = gold_total - value
                                        flag_ids.append(item_id)

                                        ### Update player_readonly_data

                                        query = {"session_id": session_id}
                                        update = {}
                                        update['player_readonly_data'] = json.dumps(player_readonly_data)

                                        projection = 'session_id player_readonly_data'
                                        response = await accounts_cache.update_account(req, ws, send_message, query, update, projection)
                                        if response is not None:
                                            response["request_id"] = req['request_id']
                                            await send_message(response, ws)  # send response
                                            valid = True
                                        else:
                                            error_response["error"] = "There was a problem updating the account."
                                    else:
                                        error_response["error"] = "You already own that flag.\nYou do not need to buy it again."
                                else:
                                    error_response["error"] = "Could not buy.\nYou do not have enough gold."
                            else:
                                error_response["error"] = "Value not valid."
                        else:
                            error_response["error"] = "Item id out of bounds"
                    else:
                        error_response["error"] = "Item id not valid."
                else:
                    error_response["error"] = "Item type not valid."
            else:
                error_response["error"] = "Game not valid."
        else:
            error_response["error"] = "Session is expired."
    else:
        error_response["error"] = "No account found."

    ### Envia error al cliente.

    if not valid:
        await send_message(error_response, ws)

    

    

File: account-server-golang/server_api/BUY_WITH_GOLD.go
package server_api

import (
	"account-server/database"
	"account-server/resource_manager"
	"account-server/utils"
	"encoding/json"
	"errors"
	"time"

	"go.mongodb.org/mongo-driver/bson"
	"golang.org/x/exp/slices"
)

type request_BUY_WITH_GOLD struct {
	Request_id string
	Session_id string
	Game_name  string
	Item_type  string
	Item_id    uint
}

// BUY_WITH_GOLD
// 1. El cliente envia BUY_WITH_GOLD(session_id) al servidor.
// 2. El servidor comprueba que existe esa session_id y no esta caducada.
// 3. El servidor responde con el resource o con error.
func BUY_WITH_GOLD(message []byte, sendResponse func([]byte)) error {

	var r request_BUY_WITH_GOLD

	// Comprueba que estan todos los parametros requeridos.
	{
		var m map[string]any
		err := json.Unmarshal(message, &m)
		if err != nil {
			return err
		}
		if !utils.ContainsAllKeys(m, []string{"request_id", "session_id", "game_name", "item_type", "item_id"}) {
			err = errors.New("BUY_WITH_GOLD bad input parameters")
			return err
		}

		if _, ok := m["require_test_instance"]; ok { // Evita que las requests, que indican require_test_instance, se ejecuten en una instancia del account server que no es de test.
			if !database.IsTestInstance() {
				return errors.New("require_test_instance")
			}
		}
	}

	err := json.Unmarshal(message, &r)
	if err != nil {
		return err
	}

	var account dict

	// Comprueba si existe una cuenta con esa session_id
	{
		target := []string{"session_id", r.Session_id}
		responseDict, err := database.ReadAccount(target, "guest_account_id session_expire player_readonly_data")
		if err != nil {
			return err
		}
		account = responseDict
	}

	responseDict := dict{}

	if account != nil {

		sessionExpired := false

		if account["guest_account_id"] != nil {
			// Para las cuentas guest no se aplica session_expire
		} else {
			// Comprueba que la session no esta caducada
			expireTimeStr := account["session_expire"].(string) // Sera "" si session_expire no esta en account
			if expireTimeStr == "" {
				panic(0)
			}
			expireTime, err := time.Parse(time.RFC3339, expireTimeStr)
			if err != nil {
				panic(0)
			}
			sessionExpired = expireTime.Before(time.Now())
		}

		if !sessionExpired {

			if r.Game_name == "tanks" {

				game_parameters := resource_manager.GetGame("tanks")["GAME_PARAMETERS"].(dict)

				if r.Item_type == "skin" { // SKIN
					max_number_of_skins := uint(game_parameters["MAX_NUMBER_OF_SKINS"].(float64))
					if r.Item_id < max_number_of_skins {
						var rod dict
						err := json.Unmarshal([]byte(account["player_readonly_data"].(string)), &rod)
						if err != nil {
							return err
						}
						skin_ids := rod["skin_ids"].([]any)
						gold_total := rod["gold"].(float64)
						value := game_parameters["SKIN_PRICES"].([]any)[r.Item_id].(float64)
						if value <= gold_total {
							if slices.IndexFunc(skin_ids, func(skin_id any) bool { return uint(skin_id.(float64)) == r.Item_id }) < 0 {
								rod["gold"] = gold_total - value
								rod["skin_ids"] = append(skin_ids, r.Item_id)

								target := []string{"session_id", r.Session_id}
								update := bson.D{bson.E{}}
								{
									bytes, err := json.Marshal(rod)
									if err != nil {
										return err
									}
									update[0] = bson.E{Key: "player_readonly_data", Value: string(bytes)}
								}
								// projection := "session_id player_readonly_data"

								err := database.UpdateAccount(target, update)
								if err == nil {
									responseDict["session_id"] = r.Session_id
									responseDict["player_readonly_data"] = update[0].Value
								} else {
									responseDict["error"] = "There was a problem updating the account."
								}
							} else {
								responseDict["error"] = "You already own that item.\nYou do not need to buy it again."
							}
						} else {
							responseDict["error"] = "Could not buy.\nYou do not have enough gold."
						}
					} else {
						responseDict["error"] = "Item id out of bounds"
					}
				} else if r.Item_type == "flag" { // FLAG
					max_number_of_flags := uint(game_parameters["MAX_NUMBER_OF_FLAGS"].(float64))
					if r.Item_id < max_number_of_flags {
						var rod dict
						err := json.Unmarshal([]byte(account["player_readonly_data"].(string)), &rod)
						if err != nil {
							return err
						}
						flag_ids := rod["flag_ids"].([]any)
						gold_total := rod["gold"].(float64)
						value := game_parameters["FLAG_PRICE"].(float64)
						if value <= gold_total {
							if slices.IndexFunc(flag_ids, func(flag_id any) bool { return uint(flag_id.(float64)) == r.Item_id }) < 0 {
								rod["gold"] = gold_total - value
								rod["flag_ids"] = append(flag_ids, r.Item_id)

								target := []string{"session_id", r.Session_id}
								update := bson.D{bson.E{}}
								{
									bytes, err := json.Marshal(rod)
									if err != nil {
										return err
									}
									update[0] = bson.E{Key: "player_readonly_data", Value: string(bytes)}
								}
								// projection := "session_id player_readonly_data"

								err := database.UpdateAccount(target, update)
								if err == nil {
									responseDict["session_id"] = r.Session_id
									responseDict["player_readonly_data"] = update[0].Value
								} else {
									responseDict["error"] = "There was a problem updating the account."
								}
							} else {
								responseDict["error"] = "You already own that item.\nYou do not need to buy it again."
							}
						} else {
							responseDict["error"] = "Could not buy.\nYou do not have enough gold."
						}
					} else {
						responseDict["error"] = "Item id out of bounds"
					}
				} else {
					responseDict["error"] = "Item type not valid."
				}
			} else {
				responseDict["error"] = "Game not valid."
			}
		} else {
			responseDict["error"] = "Session is expired."
		}
	} else {
		responseDict["error"] = "No account found with this session_id."
	}

	// Envia la respuesta

	responseDict["request_id"] = r.Request_id
	delete(responseDict, "guest_account_id")

	response, err := json.Marshal(responseDict)
	if err != nil {
		return err
	}
	sendResponse(response)

	return nil
}
